### PR TITLE
Init get_weather: gets Date and Location and calls API for weather-fo…

### PR DESCRIPTION
**Pull request by arno-hartmann**
…recast, writes that into DynamoDB

Please check what I have created!
Grateful for comments and suggestions!

### CODE REVIEW

File: formatting_Utils.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from decimal import *
```
Comment (fabianschmauder) [inline formatting_Utils.py:1]: please rename the file to 
formating_utils.py
;)

File: get_weather_Utils.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
+from decimal import *
```
Comment (fabianschmauder) [inline get_weather_Utils.py:1]: same file  renaming

File: get_weather.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import formatting_Utils
+import get_weather_Utils
+
+# Gets today's date and makes it a string
+today = date.today()
+today_str = formatting_Utils.maketimestring(today)
+
+#Gets the city Name and ID from Database (here 0), will be called in a while/for to get all cities
+city = get_weather_Utils.get_location(0)[0]
+city_id = get_weather_Utils.get_location(0)[1]
```
Comment (fabianschmauder) [inline get_weather.py]: why do you call get_location twice. Better save it to a variable

File: get_weather.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from decimal import *
+from datetime import date, timedelta, datetime
+import formatting_Utils
+import get_weather_Utils
+
```
Comment (fabianschmauder) [inline get_weather.py:6]: i think you do not need so many comments

File: get_weather.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+# Gets today's date and makes it a string
+today = date.today()
+today_str = formatting_Utils.maketimestring(today)
+
+#Gets the city Name and ID from Database (here 0), will be called in a while/for to get all cities
+city = get_weather_Utils.get_location(0)[0]
+city_id = get_weather_Utils.get_location(0)[1]
+
+# makes URL for calling the API for weather in City for the day that is given
+api_url = get_weather_Utils.make_url_api(city_id, today_str)
```
Comment (fabianschmauder) [inline get_weather.py]:
```suggestion
api_url = get_weather_Utils.determine_api_url(city_id, today_str)
```

File: get_weather_Utils.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import boto3
+import requests
+from datetime import date, timedelta, datetime
+import formatting_Utils
+
+
+
+def make_url_api(location_id, day_str):
```
Comment (fabianschmauder) [inline get_weather_Utils.py]:
```suggestion
def determine_api_url(location_id, day_str):
```

File: get_weather_Utils.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    url = "https://www.metaweather.com/api/location/" + location_id + "/"+ day_str  + "/"
+    print("API-URL aus Funktion make_url_api: ", url)
+    return url
+
+
+def get_location(number):
```
Comment (fabianschmauder) [inline get_weather_Utils.py]: number?
does every location have a number?

Comment (fabianschmauder) [reply get_weather_Utils.py]: a scan can always return the items in a different order

File: get_weather_Utils.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    dynamodb = boto3.resource('dynamodb')
+    table_destination = dynamodb.Table('destination')
+    response = table_destination.scan()
+    data = response['Items']
+    while 'LastEvaluatedKey' in response:
+        response = table_destination.scan(ExclusiveStartKey=response['LastEvaluatedKey'])
+        data.extend(response['Items'])
+    city = response['Items'][number]['city']
```
Comment (fabianschmauder) [inline get_weather_Utils.py]: could lead to an error when db is empty

File: get_weather_Utils.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+def make_url_api(location_id, day_str):
+    url = "https://www.metaweather.com/api/location/" + location_id + "/"+ day_str  + "/"
+    print("API-URL aus Funktion make_url_api: ", url)
+    return url
+
+
+def get_location(number):
+    dynamodb = boto3.resource('dynamodb')
+    table_destination = dynamodb.Table('destination')
+    response = table_destination.scan()
```
Comment (fabianschmauder) [inline get_weather_Utils.py:17]: scan is very expensive. 
Currently you load all data that you have. 
would be better to search for a specific location

File: get_weather_Utils.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    data = response['Items']
+    while 'LastEvaluatedKey' in response:
+        response = table_destination.scan(ExclusiveStartKey=response['LastEvaluatedKey'])
+        data.extend(response['Items'])
+    city = response['Items'][number]['city']
+    city_id = response['Items'][number]['city_id']
+    print(city, city_id)
+    return [city, city_id]
```
Comment (fabianschmauder) [inline get_weather_Utils.py]: why array?
i would return a dict. 
it is very implicit that the first is a city and the second item in the list is an id

File: get_weather_Utils.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+
+
+def write_weather_dynamoDB(api_url, location):
+    url = requests.get(api_url)
+    response = url.json()
+
+    dynamodb = boto3.resource('dynamodb')
```
Comment (fabianschmauder) [inline get_weather_Utils.py]: you use this multiple times in different functions. maybe move it up

File: get_weather_Utils.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    table_weather = dynamodb.Table('weather')
+
+
+    counter = 0
+    for item in response:
+        
+        #city in DB
+        # date = applicable_date
+        report = counter
+        id = response[counter]['id']
```
Comment (fabianschmauder) [inline get_weather_Utils.py]: why you are not using the item directly? 
```suggestion
        id = item['id']
```
its the same for the other parameters

File: get_weather_Utils.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        max_temp = formatting_Utils.make_decimal(response[counter]['max_temp'])
+        the_temp = formatting_Utils.make_decimal(response[counter]['the_temp'])
+        wind_speed = formatting_Utils.make_decimal(response[counter]['wind_speed'])
+        wind_direction = formatting_Utils.make_decimal(response[counter]['wind_direction'])
+        air_pressure = formatting_Utils.make_decimal(response[counter]['air_pressure'])
+        humidity = response[counter]['humidity']
+        visibility = formatting_Utils.make_decimal(response[counter]['visibility'])
+        predictability = response[counter]['predictability']
+        
+        if formatting_Utils.is_date_recent(created):
```
Comment (fabianschmauder) [inline get_weather_Utils.py]: that is the first think i would do in the loop
```suggestion
        if not formatting_Utils.is_date_recent(created):
            continue
```

File: get_weather_Utils.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+            dict = {
+                'city'  : location,
+                'date'  : counter,
+                'id'    : id,
```
Comment (fabianschmauder) [inline get_weather_Utils.py]:
```suggestion
                'id'    : item['id'],
```
than you do not need a variable

**Review by fabianschmauder — COMMENTED** [review]
a few comments 😉

File: formatting_Utils.py (hunk 1 above)
Comment (arno-hartmann) [reply formatting_Utils.py:1]: Done

File: get_weather_Utils.py (hunk 2 above)
Comment (arno-hartmann) [reply get_weather_Utils.py:1]: Done

File: get_weather.py (hunk 4 above)
Comment (arno-hartmann) [reply get_weather.py:6]: Alright. Was just for my reviewers to get into it easier.

File: get_weather.py (hunk 3 above)
Comment (arno-hartmann) [reply get_weather.py]: Done. Getting a list of dicts back! 
Great Idea!

File: get_weather.py (hunk 5 above)
Comment (arno-hartmann) [reply get_weather.py]: 👍🏼

File: get_weather_Utils.py (hunk 6 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: 👍🏼

File: get_weather_Utils.py (hunk 7 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: Implemented: get a list of dicts returned. Works like a charm!
Thank you!

File: get_weather_Utils.py (hunk 9 above)
Comment (arno-hartmann) [reply get_weather_Utils.py:17]: 🤔 Will look into that. 
It's a small Table of "city" and "city_id". Want to get weather for all of them.

File: get_weather_Utils.py (hunk 10 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: Resolved with list of dicts

File: get_weather_Utils.py (hunk 12 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: Done 👍🏼

File: get_weather_Utils.py (hunk 14 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: awesome! 👍🏼

File: get_weather_Utils.py (hunk 13 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: Done. 👍🏼

File: get_weather_Utils.py (hunk 11 above)
Comment (arno-hartmann) [reply get_weather_Utils.py]: Done

File: formatting_Utils.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    return timestamp
+
+def is_date_recent(data):
+    day = date.today()
+    yesterday = day - timedelta(days = 2)
+    t = datetime.strptime(data, "%Y-%m-%dT%H:%M:%S.%fZ")
+    if t.date() > yesterday:
+        return True
+    else:
+        return False
```
Comment (fabianschmauder) [inline formatting_Utils.py]:
```suggestion
    return t.date() > yesterday
```
;)

**Review by fabianschmauder — COMMENTED** [review]
:)

**Pull request merged**